### DISCUSSION TITLE
Copy and paste inside textual

### DISCUSSION THREAD

**Discussion by novak-services** [Textualize/textual#2190] [Q&A] [answered]
I recognized, that when running the textual demo app, I can not select text or even do copy and paste (maybe paste, but definitively not copy). Tested inside xfce4 terminal and also a MinGW terminal (from Windows 10 via SSH to Linux).

I like and appreciate the idea of having a modern interface based on textual terminals very much, particularly because it would give a lot of interaction possibilities. But selecting and copying text is so basic, that I don't want to loose it in the process.

My questions: 

- Is there currently some method/way to mitigate this?
- would it be possible to add some features in the future, that copying (and pasting) is available again (at least as option)?

Any answers and suggestions welcome, since I already planned to integrate textual in my own projects, but this feature is definitively missing here.

**Comment by davep** [comment, marked as answer]
As you'll have seen from the issue you opened too (best to only do a discussion or an issue; not both), there is an FAQ that covers this: https://github.com/Textualize/textual/blob/main/FAQ.md#how-can-i-select-and-copy-text-in-a-textual-app

**Reply by novak-services** [reply]
Thank you very much. This did help.

I am not used to discussions at github, sorry for opening both. The system basically motivated me to open both at a time. As much I saw, the ticket was already closed. 

Somehow, I did not see the FAQ before. But I searched the discussions. Now other people may also find the FAQ.

**Reply by davep** [reply]
No worries. Glad to hear it helped. 🙂

**Reply by mwmeyer** [reply]
link above 404ing, its https://textual.textualize.io/FAQ/ now
Screenshot for posterity
<img width="949" alt="image" src="https://github.com/Textualize/textual/assets/1466940/40e95e6c-e5d2-48fa-921b-d2ee019de908">

**Comment by etse1288** [comment]
It seems like it is not working when shift-selecting the textual app content from within visual studio code's terminal?

**Reply by TomJGooding** [reply]
Never ran Textual in VSCode before but seems to work for me testing with this: https://textual.textualize.io/widgets/static/#example

Could you run `textual diagnose` and paste the output here in case it offers some clues?

**Reply by davep** [reply]
Testing this with VSCode on macOS, I can't shift-select in the Textual demo in the terminal, but neither can I do the same in Midnight Commander running in the terminal, or htop, or some other application-mode programs.

Perhaps I'm mistaken but this seems to be an aspect of the VSCode terminal.

**Reply by etse1288** [reply]
I can shift-drag and copy and paste with `htop` running in a terminal inside vscode
![textual1](https://user-images.githubusercontent.com/103478851/232202673-4dad7c8f-d822-4c8a-914e-e75a5c2d8920.gif)

**Reply by davep** [reply]
Are you running `htop` in that terminal with mouse support? Again, testing here on macOS; if I run `htop` in a `code` terminal with mouse support enabled I can't select with the mouse; on the other hand if I run `htop --no-mouse` I *can* select with the mouse.

**Reply by etse1288** [reply]
I *can* shift-select with `htop`, `tmux`, but not with `python -m textual`. Thanks

**Reply by giohappy** [reply]
Same for me, on Windows, cooy/paste not working inside VS terminal.

**Comment by etse1288** [comment]
# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.19.1 |
| Rich    | 13.3.4 |

## Python

| Name           | Value                                            |
|----------------|--------------------------------------------------|
| Version        | 3.7.4                                            |
| Implementation | CPython                                          |
| Compiler       | GCC 4.7.2                                        |
| Executable     | /home/etse/.virtualenvs/dev1/bin/python |

## Operating System

| Name    | Value                                                   |
|---------|---------------------------------------------------------|
| System  | Linux                                                   |
| Release | 4.12.14-122.136-default                                 |
| Version | #1 SMP Wed Oct 5 09:01:28 UTC 2022 (253e49e/lp-70a9340) |

## Terminal

| Name                 | Value           |
|----------------------|-----------------|
| Terminal Application | vscode (1.77.1) |
| TERM                 | xterm-256color  |
| COLORTERM            | truecolor       |
| FORCE_COLOR          | *Not set*       |
| NO_COLOR             | *Not set*       |

## Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=174, height=75 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 174                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 75                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |

**Reply by etse1288** [reply]
![textual](https://user-images.githubusercontent.com/103478851/232171733-e9f16cbd-c546-4476-9b14-af44112ef797.gif)

**Comment by Forgen** [comment]
Two things - first the above FAQ link changed to https://github.com/Textualize/textual/blob/main/docs/FAQ.md.

Second, the FAQ approaches do not work in the web browser (with `textual serve`). Shift-click and drag highlights momentarily, but the highlight vanishes as soon as the mouse button is released. Cannot copy text.

**Reply by slanos** [reply]
Looks like it's a known issue on `textual serve`: https://github.com/Textualize/textual-serve/issues/12

I think for now the easiest thing would be creating your own element that wraps `Text` components, and manually causes the copy to happen when clicking.

**Comment by mendhak** [comment]
Does anyone have any hints as to how text could be selected for the new Android Linux Terminal?

**Comment by deadgrassllc-ux** [comment]
~So it sounds like Textual does not support copying text.  That's pretty unfortunate.  There are many scenarios in which the 'just ignore textual and use the terminal's copy functionality' doesn't work well.  Like for example if you don't want to copy a bunch of Textual UI along with the text.~

Looks like it's just whatever component I'm using does not support copy.

Link to FAQ was broken: https://textual.textualize.io/FAQ/
